watch/funcs_test.go

```diff
@@ -1,137 +1,170 @@
-package watch
+package watch_test
 
 import (
-	"os"
+	"encoding/json"
+	"errors"
+	"sync"
 	"testing"
 	"time"
 
+	"github.com/hashicorp/consul/agent"
 	consulapi "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/watch"
 )
 
-var consulAddr string
-
-func init() {
-	consulAddr = os.Getenv("CONSUL_ADDR")
-}
+var errNoContent = errors.New("no content")
+var errBadContent = errors.New("bad content")
 
 func TestKeyWatch(t *testing.T) {
-	if consulAddr == "" {
-		t.Skip()
-	}
+	a := agent.NewTestAgent(t.Name(), ``)
+	defer a.Shutdown()
+
+	invoke := make(chan error)
 	plan := mustParse(t, `{"type":"key", "key":"foo/bar/baz"}`)
-	invoke := 0
 	plan.Handler = func(idx uint64, raw interface{}) {
-		if invoke == 0 {
+		invoke <- func() error {
 			if raw == nil {
-				return
+				return errNoContent
 			}
 			v, ok := raw.(*consulapi.KVPair)
-			if !ok || v == nil || string(v.Value) != "test" {
-				t.Fatalf("Bad: %#v", raw)
+			if !ok {
+				return errBadContent
 			}
-			invoke++
-		}
+			if v == nil {
+				return errNoContent
+			}
+			if string(v.Value) != "test" {
+				return errBadContent
+			}
+			return nil
+		}()
 	}
 
+	var wg sync.WaitGroup
+	wg.Add(1)
 	go func() {
-		defer plan.Stop()
-		time.Sleep(20 * time.Millisecond)
+		defer wg.Done()
+		kv := a.Client().KV()
 
-		kv := plan.client.KV()
+		time.Sleep(20 * time.Millisecond)
 		pair := &consulapi.KVPair{
 			Key:   "foo/bar/baz",
 			Value: []byte("test"),
 		}
-		_, err := kv.Put(pair, nil)
-		if err != nil {
+		if _, err := kv.Put(pair, nil); err != nil {
 			t.Fatalf("err: %v", err)
 		}
 
-		// Wait for the query to run
 		time.Sleep(20 * time.Millisecond)
```

Comment:
This sleep can be removed now?

Reply:
My concern is that when we do this we may get a single event for both raft changes instead of two. Not sure whether that is possible, though. Would have to dig through the code.